stats.py

```diff
@@ -1,187 +1,280 @@
+        if not companion_data['commander'].get('docked'):
+            out: List[ShipRet] = []
             # Set current system, not last docked
-            return ([ (str(ships[0]['id']), ship_map.get(ships[0]['name'].lower(), ships[0]['name']), ships[0].get('shipName', ''), data['lastSystem']['name'], '', str(ships[0]['value']['total'])) ] +
-                    [ (str(ship['id']), ship_map.get(ship['name'].lower(), ship['name']), ship.get('shipName', ''), ship['starsystem']['name'], ship['station']['name'], str(ship['value']['total'])) for ship in ships[1:] if ship])
+            out.append(ShipRet(
+                id=str(ships[0]['id']),
+                type=ship_map.get(ships[0]['name'].lower(), ships[0]['name']),
+                name=str(ships[0].get('shipName', '')),
+                system=companion_data['lastSystem']['name'],
+                station='',
+                value=str(ships[0]['value']['total'])
+            ))
+            out.extend(
+                ShipRet(
+                    id=str(ship['id']),
+                    type=ship_map.get(ship['name'].lower(), ship['name']),
+                    name=ship.get('shipName', ''),
+                    system=ship['starsystem']['name'],
+                    station=ship['station']['name'],
+                    value=str(ship['value']['total'])
+                ) for ship in ships[1:] if ship
+            )
+
+            return out
+
+    return [
+        ShipRet(
+            id=str(ship['id']),
+            type=ship_map.get(ship['name'].lower(), ship['name']),
+            name=ship.get('shipName', ''),
+            system=ship['starsystem']['name'],
+            station=ship['station']['name'],
+            value=str(ship['value']['total'])
+        ) for ship in ships if ship is not None
+    ]
+
 
-    return [ (str(ship['id']), ship_map.get(ship['name'].lower(), ship['name']), ship.get('shipName', ''), ship['starsystem']['name'], ship['station']['name'], str(ship['value']['total'])) for ship in ships if ship]
+def export_ships(companion_data: Dict[str, Any], filename: AnyStr) -> None:
+    """
+    Export the current ships to a CSV file.
 
-def export_ships(data, filename):
-    h = csv.writer(open(filename, 'w'))
-    h.writerow(['Id', 'Ship', 'Name', 'System', 'Station', 'Value'])
-    for thing in ships(data):
-        h.writerow([x for x in thing])
+    :param companion_data: Data from which to generate the ship list
+    :param filename: The target file
+    """
+    with open(filename, 'w') as f:
+        h = csv.writer(f)
+        h.writerow(['Id', 'Ship', 'Name', 'System', 'Station', 'Value'])
+        for thing in ships(companion_data):
+            h.writerow(list(thing))
 
 
 class StatsDialog():
+    """Status dialog containing all of the current cmdr's stats."""
 
-    def __init__(self, app):
-        self.parent = app.w
+    def __init__(self, app: 'AppWindow') -> None:
+        self.parent: tk.Tk = app.w
```

Comment:
Couldn't this just pass in app.w given 'app' isn't otherwise used ?  I'm done with EDMarketConnector.py now, so you can make the requisite change that side too.